addon-test-support/browser-state.js

```diff
@@ -0,0 +1,119 @@
+let initialRequireEntries;
+let initialScriptTags;
+let initialLinkTags;
+
+/**
+ * Determines whether an array contains the provided item.
+ *
+ * @param {Array} array
+ * @param {Any} item
+ * @return {Boolean}
+ */
+function has(array, item) {
+  return array.indexOf(item) !== -1;
+}
+
+/**
+ * Removes a DOM Node from the document.
+ *
+ * @param {Node} node
+ * @return {Void}
+ */
+function removeNode(node) {
+  node.parentNode.removeChild(node);
+}
+
+/**
+ * Converts an iterable object into an actual Array.
+ *
+ * @param {Iterable} iterable
+ * @return {Array}
+ */
+function toArray(iterable) {
+  return Array.prototype.slice.call(iterable)
+}
+
+/**
+ * Returns all of the HTML Elements matching a given selector as an array.
+ *
+ * @param {String} selector
+ * @return {Array<HTMLElement>}
+ */
+function getAll(selector) {
+  const htmlCollection = document.querySelectorAll(selector);
+  return toArray(htmlCollection);
+}
+
+/**
+ * Deletes an entry from require's list of modules.
+ *
+ * @param {String} entry
+ * @return {Void}
+ */
+function resetRequireEntry(entry) {
+  delete requirejs.entries[entry];
+}
+
+/**
+ * Compares two arrays, if they're different, invokes a callback for each
+ * entry that does not appear in the initial array.
+ *
+ * @param {Array} initial
+ * @param {Array} current
+ * @param {Function} diffHandler
+ * @return {Void}
+ */
+function compareAndIterate(initial, current, diffHandler) {
+  if (initial.length < current.length) {
+    for (let i = 0; i < current.length; i++) {
+      let entry = current[i];
+      if (!has(initial, entry)) {
+        diffHandler(entry);
+      }
+    }
+  }
+}
+
+/**
+ * Gets the current browser state including scripts, links, and require modules.
+ *
+ * @return {Object}
+ */
+export function getBrowserState() {
+  return {
+    requireEntries: Object.keys(requirejs.entries),
+    scripts: getAll('script'),
+    links: getAll('link')
+  };
+}
+
+/**
+ * Caches the current browser state with regards to links, scripts, and JS
+ * modules currently present.
+ *
+ * @return {Void}
+ */
+export function cacheBrowserState() {
+  ({
+    requireEntries: initialRequireEntries,
+    scripts: initialScriptTags,
+    links: initialLinkTags
+  } = getBrowserState());
+}
+
+/**
+ * Restores the browser state to the previous cached value with regards to
+ * links, scripts, and JS modules.
+ *
+ * @return {Void}
+ */
+export function resetBrowserState() {
```

Comment:
@nathanhammond:

These are test helpers that are generally ran *after* app teardown, so we generally do not have to worry about services / controllers holding onto state (since they would be destroyed).  

We do still need to document that any module level side effects of non-lazy engine modules would have to be manually cleaned up though.